src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder_ListPatterns.cs

```diff
@@ -55,17 +56,19 @@ private Tests MakeTestsAndBindingsForListPattern(BoundDagTemp input, BoundListPa
 
                         if (slice.Pattern is BoundPattern slicePattern)
                         {
-                            Debug.Assert(slice.IndexerAccess is not null);
-                            Debug.Assert(index <= 0);
-                            Debug.Assert(slice.ReceiverPlaceholder is not null);
-                            Debug.Assert(slice.ArgumentPlaceholder is not null);
+                            // We faced a slice pattern with a declaration pattern of the same narrowed type inside, e.g. `[<pattern>, ..var between, <pattern>]`
+                            // This pattern always evaluates to true, but causes slicing operation and even may allocate (e.g. when slicing an array or string)
+                            // Thus it makes sense to hold it and "evaluate" after all other patterns in the list matched
+                            if (slicePattern.Kind == BoundKind.DeclarationPattern && slicePattern.InputType.Equals(slicePattern.NarrowedType, TypeCompareKind.AllNullableIgnoreOptions))
```

Comment:
>TypeCompareKind.AllNullableIgnoreOptions [](http://example.com/codeflow?start=142&length=40)

Do we have tests covering the comparison kind?

Reply:
Need to resolve https://github.com/dotnet/roslyn/pull/70615#discussion_r1539756164 before working on this one